File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/ui/client/MyShopUiClient.java
package com.mycompany.myshop.testkit.driver.adapter.ui.client;

import com.microsoft.playwright.*;
import com.mycompany.myshop.testkit.driver.adapter.ui.client.pages.HomePage;
import com.mycompany.myshop.testkit.common.Closer;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.playwright.PageClient;
import org.springframework.http.HttpStatus;

public class MyShopUiClient implements AutoCloseable {
    private static final String CONTENT_TYPE = "content-type";
    private static final String TEXT_HTML = "text/html";
    private static final String HTML_OPENING_TAG = "<html";
    private static final String HTML_CLOSING_TAG = "</html>";

    private final String baseUrl;
    private final BrowserContext context;
    private final Page page;
    private final PageClient pageClient;
    private final HomePage homePage;

    private Response response;

    public MyShopUiClient(String baseUrl, Browser browser) {
        this.baseUrl = baseUrl;

        // Create isolated browser context for this test instance
        var contextOptions = new Browser.NewContextOptions()
                .setViewportSize(1920, 1080)
                // Ensure complete isolation between parallel tests
                .setStorageStatePath(null);

        this.context = browser.newContext(contextOptions);

        // Each test gets its own page
        this.page = context.newPage();

        this.pageClient = new PageClient(page);
        this.homePage = new HomePage(pageClient);
    }

    public HomePage openHomePage() {
        response = page.navigate(baseUrl);
        return homePage;
    }

    public boolean isStatusOk() {
        return response.status() == HttpStatus.OK.value();
    }

    public boolean isPageLoaded() {
        if (response == null || response.status() != HttpStatus.OK.value()) {
            return false;
        }

        var contentType = response.headers().get(CONTENT_TYPE);
        if (contentType == null || !contentType.startsWith(TEXT_HTML)) {
            return false;
        }

        var pageContent = page.content();
        return pageContent != null && pageContent.contains(HTML_OPENING_TAG) && pageContent.contains(HTML_CLOSING_TAG);
    }

    @Override
    public void close() {
        Closer.close(page);
        Closer.close(context);
        // Don't close browser - it's shared and managed by test lifecycle infrastructure
    }
}

File: system-test/dotnet/Driver.Adapter/Ui/Client/MyShopUiClient.cs
using Microsoft.Playwright;

using Driver.Adapter.Ui.Client.Pages;

using Driver.Adapter.Shared.Client.Playwright;

using System.Net;

using PlaywrightGateway = Driver.Adapter.Shared.Client.Playwright.PageClient;



namespace Driver.Adapter.Ui.Client;



public class MyShopUiClient : IAsyncDisposable

{

    // Default: headless mode (browser not visible)

    // To see browser during debugging, set: HEADED=true or PLAYWRIGHT_HEADED=true

    private static readonly bool IsHeadless = Environment.GetEnvironmentVariable("HEADED") != "true";



    private const string ContentType = "content-type";

    private const string TextHtml = "text/html";

    private const string HtmlOpeningTag = "<html";

    private const string HtmlClosingTag = "</html>";



    private readonly string _baseUrl;

    private readonly IPlaywright _playwright;

    private readonly IBrowser _browser;

    private readonly IBrowserContext _context;

    private readonly IPage _page;

    private readonly HomePage _homePage;



    private IResponse? _response;



    private MyShopUiClient(string baseUrl, IPlaywright playwright, IBrowser browser, IBrowserContext context, IPage page, HomePage homePage)

    {

        _baseUrl = baseUrl;

        _playwright = playwright;

        _browser = browser;

        _context = context;

        _page = page;

        _homePage = homePage;

    }



    public static async Task<MyShopUiClient> CreateAsync(string baseUrl)

    {

        var playwright = await Playwright.CreateAsync();

        var browser = await playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions { Headless = IsHeadless });



        // Create isolated browser context with specific configuration

        var contextOptions = new BrowserNewContextOptions

        {

            ViewportSize = new ViewportSize { Width = 1920, Height = 1080 },

            StorageStatePath = null // Ensure complete isolation between parallel tests

        };

        var context = await browser.NewContextAsync(contextOptions);



        // Each test gets its own page

        var page = await context.NewPageAsync();

        var pageClient = new PlaywrightGateway(page, baseUrl);

        var homePage = new HomePage(pageClient);



        return new MyShopUiClient(baseUrl, playwright, browser, context, page, homePage);

    }



    public async Task<HomePage> OpenHomePageAsync()

    {

        _response = await _page.GotoAsync(_baseUrl);

        return _homePage;

    }



    public bool IsStatusOk()

    {

        return _response?.Status == ((int)HttpStatusCode.OK);

    }



    public async Task<bool> IsPageLoadedAsync()

    {

        if (_response == null || _response.Status != ((int)HttpStatusCode.OK))

        {

            return false;

        }



        var contentType = _response.Headers.ContainsKey(ContentType) ? _response.Headers[ContentType] : null;

        if (contentType == null || !contentType.StartsWith(TextHtml))

        {

            return false;

        }



        var pageContent = await _page.ContentAsync();

        return pageContent != null &&

               pageContent.Contains(HtmlOpeningTag) &&

               pageContent.Contains(HtmlClosingTag);

    }



    public async ValueTask DisposeAsync()

    {

        if (_page != null)

            await _page.CloseAsync();

        if (_context != null)

            await _context.CloseAsync();

        if (_browser != null)

            await _browser.CloseAsync();



        _playwright?.Dispose();

    }

}





